### PR TITLE
New Mapping Tool: dmm2tgm.py

### PR DESCRIPTION
**Pull request by RemieRichards**
## What is dmm2tgm.py?

This converts the internal syntax of a .dmm file (<B>Example 1</B>) to a format that can be read better by humans (<B>Example 2</B>) but the best part is that this should reduce mapping conflicts.
Mapping conflicts should now only occur when two or more people touch the EXACT SAME ITEM or EXACT SAME VARIABLE on map instances, where as at the moment it conflicts arise the moment people even remotely touch two turfs near each other.

<B>Example 1 (Normal .dmm syntax)</B>

```
"aab" = (/obj/docking_port/stationary{dheight = 9;dir = 2;dwidth = 5;height = 22; id = "syndicate_n";name = "north of station";turf_type = /turf/space;width = 18},/turf/space,/area/space)
```

<B>Example 2 (dmm2tgm.py processed .dmm syntax)</B>

```
"aab" = (
/obj/docking_port/stationary{
        dheight = 9;
         dir = 2;
         dwidth = 5;
         height = 22;
         id = "syndicate_n";
         name = "north of station";
         turf_type = /turf/space;
         width = 18
        },
/turf/space,
/area/space)
```
## Disclaimer:

This tool is a bit clunky to use at the moment, you must specify an input map (from within the same directory as the tool) and an output map (from within the same directory as the tool, however if you leave the output map variable empty it will simply overwrite the input map file

Eventually this tool will be run automatically by the other mapping tools (map merger, etc.) so you won't even need to worry about it, but for now it's experimental.

No maps were edited as this would almost CERTAINLY conflict with every map PR currently up.

The tool will NOT run on maps containing this header comment:
`//MAP CONVERTED BY dmm2tgm.py THIS HEADER COMMENT PREVENTS RECONVERSION, DO NOT REMOVE`
This comment is automatically added by the tool, to prevent maps from being processed multiple times. this comment is automatically removed by byond upon saving map changes (and the map is converted back to the default format by byond) so this won't cause issues if you make a change and then re-use the tool.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Oh man, this is great, thanks @RemieRichards!

**Comment by Chnkr** [comment]
would this impact maploader performance?

**Comment by RemieRichards** [comment]
@Chnkr at this earliest instance, the maploader is not compatible with this new format, as the maploader reads line by line to find a syntax similar to byond's default syntax for .dmm files, however this tool completely restructures that syntax.

a version of the maploader could be written to handle this new syntax,  however we only actually use the maploader for awaymissions, which we currently don't load any.

the best fix would be to make the maploader not work line-by-line but instead case-by-case, correctly realising that anything between "()" is an object instance, instead of checking for this just on one line.

<B>tl;dr</B> not a hard fix, but it doesn't really need fixing given the lack of use of the maploader.

**Comment by optimumtact** [comment]
Can you rip that code out then?

**Comment by Iamgoofball** [comment]
so uh

byond can read this and shit?

i can run a map through this, plop it in my maps folder, and have it work fine?

**Comment by RemieRichards** [comment]
@Iamgoofball Byond is 100% capable of reading it. However the moment you save the map using dream maker (for example you changed your map, and saved it) Byond will reconvert it back to the normal format, But eventually this tool will be automatically ran by the other map merge tools, meaning you'll never have to worry about converting manually.

**Comment by CorruptComputer** [comment]
Why is this not byond's standard save format? :+1:

**Comment by optimumtact** [comment]
It was invented before git?

**Comment by CorruptComputer** [comment]
Probably, but updating is a thing.

**Comment by Xhuis** [comment]
kreyg
:+1:

**Comment by CorruptComputer** [comment]
http://www.byond.com/forum/?post=1896428

**Comment by RemieRichards** [comment]
@CorruptComputer a coordinate indexed system requires lots of duplication.
It would be nice if they'd just make this the default, with each object (and it's variables) on individual lines, so that git has less to complain about during conflicts.

**Comment by CorruptComputer** [comment]
Yes, but it makes it even simpler for git to handle. And the chances for conflict are almost none! You could edit the same tile and still have no conflicts (unless you edit the same line).

**Comment by RemieRichards** [comment]
@CorruptComputer which is identical to this system, provided people don't touch the same object on a turf or the same variable in an object, there are no conflicts. it's a kin to it just being a normal code file conflict (since .dmms are just .dm files with special syntax)

**Comment by Krausus** [comment]
> Mapping conflicts should now only occur when two or more people touch the EXACT SAME ITEM or EXACT SAME VARIABLE on map instances

Won't conflicts still occur if instances are created/removed? This looks like it will only affect conflicts that arise from two people editing the exact same instance.

**Comment by RemieRichards** [comment]
@Krausus I consider completely adding or completely removing an instance to to be "touching" it.
however given that this format has each instance (and each instance variable) on separate lines, it shouldn't be an issue for git to solve.

the primary issue is that the actual byond format has all the information on one line, meaning according to git (which works line by line) you've altered it all, when you've only altered a small piece.

**Comment by Krausus** [comment]
Okay, I'm using the terminology wrong. I'm referring to the, uhh... tile definitions? I'll call them that. "aaa", "aab", and so forth. Each of those is already on its own line, and should only conflict if someone touches the same definition. The bigger problem is when new definitions are created/deleted, which means changes to the huge blobs at the bottom of the map files. The ones that are so big, Git doesn't even try highlighting changes to them. Two PRs that change the definitions on the same row of the map will still conflict, even if they're on opposite sides of the map, even with this change, won't they?

**Comment by RemieRichards** [comment]
@Krausus that can't be fixed, it's a fundamental flaw of the format the maps are being saved in, it can't be changed on our end as byond to my knowledge won't read the actual map part in any other shape or structure. so yes, complete removals of "aaa"s or additions of them will still conflict, but the vast majority of our "aaa" definitions are not shared, they're for unique tile mixtures, so they won't actually change when an item is added or removed.

tl;dr it's not fixable on our end.

**Comment by CorruptComputer** [comment]
>that can't be fixed, it's a fundamental flaw of the format the maps are being saved in

@RemieRichards This is why I proposed a coordinate based system.

**Comment by RemieRichards** [comment]
@CorruptComputer your system has other flaws, larger filesizes and map loading times from duplication and the inability to share "aaa" definitions between turfs, just look at a map at the moment, the amount of "aaa" for space is QUITE a lot, so to have to replace every single one of those with an individual defintion for (X,Y,Z) = (/area/space, /turf/space) is heavy redundancy.

**Comment by CorruptComputer** [comment]
(X,Y,Z)(X,Y,Z)(X,Y,Z)(X,Y,Z)(X,Y,Z)(X,Y,Z) = (/area/space, /turf/space) 

Have multiple coordinates on the same definition. I came up with that idea in like 2 mins, but if you would like I could probably come up with a better one.

**Comment by RemieRichards** [comment]
@CorruptComputer don't do so here, I'd recommend doing it in your byond thread, where Ter actually wants people to drum up ideas for a better format.

at the moment your proposed format (even if it took 2 minutes to formulate) just seems much worse than present to me.

**Comment by Krausus** [comment]
> yes, complete removals of "aaa"s or additions of them will still conflict, but the vast majority of our "aaa" definitions are not shared

Ah, alright. The PR description made it sound like it would prevent those, and I wasn't sure if I was missing something.

> it's not fixable on our end.

Well, theoretically, if you removed the deduplication (re-duplicated?) from the map files, and assigned every tile a fixed, unique ID, you could ensure the tile ID lumps never change, but...

> larger filesizes and map loading times

... the file sizes _would_ bloat considerably, though you could still selectively dedupe especially common tiles, such as space. I'm not actually sure about map loading times, though, because that would depend largely on how sanely BYOND processes and loads the maps.
